### PR TITLE
GEODE-7739: Address race condition in FederatingManager during listener creation

### PR DESCRIPTION
**Pull request by jhutchison**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:40]: The latch should probably be an instance of StoppableCountDownLatch so that it will abort (by throwing) if the cache is closed while threads are waiting on the latch. Otherwise, this has the potential to hang.

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:115]: Anytime you catch an InterruptedException, you should reset the interrupt bit unless the thread is locally owned and it will actually terminate when interrupted:
```
try {
  readyForEvents.await();
} catch (InterruptedException e) {
  Thread.interrupt();
  // recommended: throw new RunnableException(e); -- use an appropriate GemFire exception here
}
```
@upthewaterspout wrote an article but about this but I'm not sure where it is.

Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:115]: Also: If you don't rethrow the InterruptedException within a runtime exception, then this code probably needs to perform some sort of loop that checks a CancelCriterion. Grepping 'catch (InterruptedException' on the Geode product code (not tests) should give some good examples.

File: geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java:52]: Same problems as described for ManagementCacheListener apply here.

**Review by kirklund — CHANGES_REQUESTED** [review]

**Comment by jhutchison** [comment]
@kirklund thanks for input-  we're looking into these

File: geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:495]: new dead code. please remove... No new todo's please submit a bug.

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:48]: new dead code.

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:65]: new dead code.

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:82]: new dead code.

File: geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java:65]: unused.

File: geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:436]: new dead code

File: geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java:56]: unused method

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:117]: unused method

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (mhansonp) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:108]: unused method

**Review by mhansonp — CHANGES_REQUESTED** [review]
It seems like there is some cleanup that needs to be done based on what my editor is telling me. Please take a look at these comments. if you disagree, please disregard the comments.

**Comment by jhutchison** [comment]
> It seems like there is some cleanup that needs to be done based on what my editor is telling me. Please take a look at these comments. if you disagree, please disregard the comments.

Thanks Mark-  this was just a quick push to experiment with some stuff-- will make sure none of these are in final code :)

**Review by mhansonp — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:111]: `Thread.interrupted()` only checks to see if the interrupt bit is set (returns boolean). You want to use `Thread.currentThread().interrupt()` like this:
```
try {
  readyForEvents.await();
} catch (InterruptedException e) {
  Thread.currentThread().interrupt();
  throw new RuntimeException(e);
}
```

File: geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/management/internal/NotificationCacheListener.java:56]:
```
try {
  readyForEvents.await();
} catch (InterruptedException e) {
  Thread.currentThread().interrupt();
  throw new RuntimeException(e);
}
```

File: geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:387]: Creating a new CancelCriterion instance isn't really what is needed here. You need to `getCancelCriterion` from either the Cache or the DistributedSystem and pass that into the constructor of the classes that are using it. Then if the DS or Cache are closing, the CancelCriterion will cause the FederatingManager to cancel with the cause being the Cache or DS being closed.

If you look at #5728 (my PR with changes to FederatingManager), you'll see that the constructors are now passing in an instance of `CancelCriterion` which comes from `InternalDistributedSystem`:

```
FederatingManager(ManagementResourceRepo repo, InternalDistributedSystem system,
      SystemManagementService service, InternalCache cache, StatisticsFactory statisticsFactory,
      StatisticsClock statisticsClock, MBeanProxyFactory proxyFactory, MemberMessenger messenger,
      Supplier<ExecutorService> executorServiceSupplier) {
    this(repo, system, service, cache, statisticsFactory, statisticsClock, proxyFactory, messenger,
        system.getCancelCriterion(), executorServiceSupplier);
  }

  @VisibleForTesting
  FederatingManager(ManagementResourceRepo repo, InternalDistributedSystem system,
      SystemManagementService service, InternalCache cache, StatisticsFactory statisticsFactory,
      StatisticsClock statisticsClock, MBeanProxyFactory proxyFactory, MemberMessenger messenger,
      CancelCriterion cancelCriterion, ExecutorService executorService) {
    this(repo, system, service, cache, statisticsFactory, statisticsClock, proxyFactory, messenger,
        cancelCriterion, () -> executorService);
  }

  @VisibleForTesting
  FederatingManager(ManagementResourceRepo repo, InternalDistributedSystem system,
      SystemManagementService service, InternalCache cache, StatisticsFactory statisticsFactory,
      StatisticsClock statisticsClock, MBeanProxyFactory proxyFactory, MemberMessenger messenger,
      CancelCriterion cancelCriterion, Supplier<ExecutorService> executorServiceSupplier) {
    super(repo, system, cache, statisticsFactory, statisticsClock);
    this.service = service;
    this.proxyFactory = proxyFactory;
    this.messenger = messenger;
    this.executorServiceSupplier = executorServiceSupplier;
    lifecycleLock = new StoppableReentrantLock(cancelCriterion);
  }
```

**Review by kirklund — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/management/internal/ManagementCacheListener.java:111]: Another option which might be more correct and it matches most of the Geode code more closely:
```
boolean interrupted = false;
try {
  readyForEvents.await();
} catch (InterruptedException e) {
  interrupted = true;
  getCancelCriterion().checkCancelInProgress(ie);
  throw new RuntimeException(e);
} finally {
  if (interrupted) {
    Thread.currentThread().interrupt();
  }
}
```
The effect is that if the thread is interrupted, then we prefer Cache closed exception or DistributedSystem disconnected exception first. If neither is occurring then we throw new RuntimeException as a last resort. And then finally we reset the thread interrupt flag since we didn't actually handle the interrupt locally here by halting the Runnable that the thread is executing. The code is basically passing along a runtime exception and interrupt flag back to the calling code to let it deal with both.

File: geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java
Comment (jdeppe-pivotal) [reply geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:387]: I'm really not familiar with the usage of `CancelCriterion` so perhaps my understanding of what's required here is a bit naive... My understanding was simply that we didn't want to indefinitely block any threads `await`ing in either of the two listeners in question here. That situation would arise if something bad happened in the `FederatingManager` before it could get to calling `readyForEvents`. Hence the use of the `StoppableCountdownLatch`.

Since what's calling into these listeners is a callback using an asynchronous thread, I don't understand why it is the responsibility of these listeners to go through all of this ceremony to tell some random component that the cache is closing. Surely that component should have its own means of determining that the cache is closing? Why is it not enough for the thread to be released and throw an exception?

Comment (jhutchison) [reply geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:387]: ok, maybe it makes most sense to wait wait until 5278 is pull merged, and then adapt off that.  Are you expecting to have time to get that through in the near future?

Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:387]: I hope to have #5728 ready to merge today at the soonest or in a day or two.

Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java:387]: @jdeppe-pivotal hmm, well now you have me second-guessing the need for CancelCriterion. Maybe we can get another review before we make more changes on this PR.

**Comment by jhutchison** [comment]
code to be resubmitted in a future PR

**Pull request closed without merging**
